### PR TITLE
Implement optional contributory behaviour check

### PR DESCRIPTION
**Pull request by isislovecruft**

### CODE REVIEW

**Comment by rozbb** [comment]
I agree that this function needs to be available. The HPKE spec [requires](https://cfrg.github.io/draft-irtf-cfrg-hpke/draft-irtf-cfrg-hpke.html#section-7.1.4) this check, and the way it's [performed now](https://github.com/rozbb/rust-hpke/blob/master/src/kex/x25519.rs#L120) could be better.

I think, given this, the docs don't have to do as much justification as they currently do.

**Comment by AnomalRoil** [comment]
Totally supportive of adding this, given how people are using X25519 to do all kind of things now. 
How about also rejecting here the other points mentioned in https://cr.yp.to/ecdh.html#validate ?

And how about adding a ct check to see if the resulting shared secret is all zeroes or not?

File: src/x25519.rs [hunk 1]
```diff
@@ -177,6 +178,43 @@ impl SharedSecret {
...
+    ///
+    /// # Returns
+    ///
+    /// Returns `true` if the key exchange was contributory (good), and `false`
+    /// otherwise (can be bad for some protocols).
+    ///
+    /// [relevant]: https://tools.ietf.org/html/rfc7748#page-15
+    /// [public]: https://vnhacker.blogspot.com/2015/09/why-not-validating-curve25519-public.html
+    /// [discussions]: https://vnhacker.blogspot.com/2016/08/the-internet-of-broken-protocols.html
+    pub fn was_contributory(&self) -> bool {
```
Comment (rozbb) [inline src/x25519.rs:215]: `#[must_use]` would be appropriate here, I think

Comment (isislovecruft) [inline src/x25519.rs:215]: > `#[must_use]` seems appropriate here

Good call, thanks!

**Comment by isislovecruft** [comment]
> Totally supportive of adding this, given how people are using X25519 to do all kind of things now.
> How about also rejecting here the other points mentioned in https://cr.yp.to/ecdh.html#validate ?

That's not necessary, as those are checking that the other party's public keys are not points of small order. Any such public keys would result in the shared secret being the identity element. Thus, we only check afterwards that the shared secret is not the identity.

> And how about adding a ct check to see if the resulting shared secret is all zeroes or not?

This is a constant time check for if the resulting shared secret is the identity (all zeroes).

**Pull request merged**